src/representation_theory/monoidal_closed.lean

```diff
@@ -0,0 +1,292 @@
+  (monoidal_closed.internal_hom.flip.map (Y.ρ g)).app (op X.V)
+
+@[simp] lemma post_comp_apply :
+  post_comp X Y g = (monoidal_closed.internal_hom.flip.map (Y.ρ g)).app (op X.V) := rfl
+
+@[simp] lemma post_comp_mul :
+  post_comp X Y (g * g') = post_comp X Y g' ≫ post_comp X Y g := by simp
+
+lemma pre_post_comp_comm :
+  pre_comp X Y g ≫ post_comp X Y g' = post_comp X Y g' ≫ pre_comp X Y g := by simp
+
+/-- The internal hom of `X Y : Action V G`, i.e. the natural G-action on `[X, Y]` -/
+@[protected] def ihom (X Y : Action V G) : Action V G :=
+⟨(ihom X.V).obj Y.V,
+begin
+  refine ⟨λ g, pre_comp X Y g ≫ post_comp X Y g, _, _⟩,
+  { simp [pre_comp_apply, post_comp_apply],
+    apply category.id_comp },
+  { intros g g',
+    rw [pre_comp_mul, post_comp_mul, End.mul_def, category.assoc, category.assoc],
+    nth_rewrite 1 ←category.assoc,
+    rw pre_post_comp_comm,
+    rw category.assoc }
+end⟩
+
+@[simp] lemma ihom_V : (ihom X Y).V = (ihom X.V).obj Y.V := rfl
+
+@[simp] lemma ihom_ρ : (ihom X Y).ρ g = pre_comp X Y g ≫ post_comp X Y g := rfl
+
+/-- `Action.ihom` as a functor -/
+def ihom_functor (X : (Action V G)ᵒᵖ) : Action V G ⥤ Action V G :=
+{ obj := ihom (unop X),
+  map := λ Y Z f, begin
+    refine ⟨(monoidal_closed.internal_hom.flip.map f.hom).app (op (unop X).V), _⟩,
+    { intro,
+      simp,
+      apply congr_arg,
+      rw [←functor.map_comp, ←functor.map_comp],
+      apply congr_arg,
+      apply f.comm },
+  end,
+  map_id' := by { intro Y, ext, simp, congr },
+  map_comp' := by { intros, apply Action.hom.ext, simp } }
+
+@[simp] lemma ihom_functor_obj (X : (Action V G)ᵒᵖ) :
+  (ihom_functor X).obj Y = ihom (unop X) Y := rfl
+
+@[simp] lemma ihom_functor_map_hom (X' : (Action V G)ᵒᵖ) (f : X ⟶ Y) :
+  ((ihom_functor X').map f).hom =
+  (monoidal_closed.internal_hom.flip.map f.hom).app (op (unop X').V) := rfl
+
+/-- Mapping of homs that is compatible with `ihom_functor` -/
+def coyoneda_map {X Y : (Action V G)ᵒᵖ} (f : X ⟶ Y) :
+  ihom_functor X ⟶ ihom_functor Y :=
+begin
+  refine ⟨_, _⟩,
+  { intro Z,
+    rw [ihom_functor_obj, ihom_functor_obj],
+    refine ⟨(monoidal_closed.internal_hom.map f.unop.hom.op).app Z.V, _⟩,
+    intro,
+    simp,
+    rw [←category.assoc, ←category.assoc],
+    apply congr_fun,
+    apply congr_arg,
+    rw [←nat_trans.comp_app, ←nat_trans.comp_app],
+    apply congr_fun,
+    apply (nat_trans.ext_iff _ _).mp,
+    rw [←functor.map_comp, ←functor.map_comp],
+    apply congr_arg,
+    rw [←op_comp, ←op_comp],
+    apply congr_arg,
+    apply (f.unop.comm (g⁻¹ : G)).symm },
+  { intros X' Y' f',
+    simp,
+    apply Action.hom.ext,
+    simp }
+end
+
+@[simp] lemma coyoneda_map_app_hom {X Y : (Action V G)ᵒᵖ} (f : X ⟶ Y) :
+  ((coyoneda_map f).app Z).hom =
+  (monoidal_closed.internal_hom.map f.unop.hom.op).app Z.V := rfl
+
+/-- The Yoneda embedding for `Action V G` built from `Action.ihom`. -/
+def coyoneda : (Action V G)ᵒᵖ ⥤ Action V G ⥤ Action V G :=
```

Comment:
I'm curious - where do you need this fully functorial version of `ihom_functor`? My impression is that `ihom_functor` should be enough, but I haven't looked at your proof in detail so maybe I'm missing something.

Reply:
Yes, `ihom_functor` should be enough. I will push another version where the final proof only refers to `ihom_functor` instead of through `coyoneda.obj`. Is the full functorial version worth keeping though?

Reply:
Once we prove the category is monoidal closed, we get the full functorial version for free as [docs#category_theory.monoidal_closed.internal_hom](https://leanprover-community.github.io/mathlib_docs/category_theory/closed/monoidal.html#category_theory.monoidal_closed.internal_hom), so I don't think there's any use in keeping `coyoneda`.